src/lib/tasks/migration_preparation_task.cpp

```diff
@@ -0,0 +1,229 @@
+    }
+  }
+
+  std::sort(hot_nodes.begin(), hot_nodes.end(),
+            [&node_hottness](const auto& a, const auto& b) { return node_hottness.at(a) < node_hottness.at(b); });
+  std::sort(cold_nodes.begin(), cold_nodes.end(),
+            [&node_hottness](const auto& a, const auto& b) { return node_hottness.at(a) < node_hottness.at(b); });
+
+  return {.imbalance = inverted_entropy(node_chunk),
+          .node_hottness = node_hottness,
+          .hot_nodes = hot_nodes,
+          .cold_nodes = cold_nodes};
+}
+
+std::vector<double> get_node_hottness(const std::vector<ChunkInfo>& chunk_infos, size_t node_count) {
+  std::vector<double> node_hottness(node_count);
+  for (const auto& chunk_info : chunk_infos) {
+    node_hottness.at(chunk_info.node) += chunk_info.hottness;
+  }
+  return scale(node_hottness);
+}
+
+bool chunk_is_completed(const Chunk& chunk, const uint32_t max_chunk_size) {
+  if (chunk.size() != max_chunk_size) return false;
+  if (chunk.has_mvcc_columns()) {
+    auto mvcc_columns = chunk.mvcc_columns();
+    for (const auto begin_cid : mvcc_columns->begin_cids) {
+      if (begin_cid == Chunk::MAX_COMMIT_ID) return false;
+    }
+  }
+  return true;
+}
+
+int get_node_id(const PolymorphicAllocator<size_t>& alloc) {
+  const auto memsource = dynamic_cast<NUMAMemoryResource*>(alloc.resource());
+  if (memsource) {
+    return memsource->get_node_id();
+  }
+  return -1;
+}
+
+std::vector<ChunkInfo> find_hot_chunks(const StorageManager& storage_manager, const std::chrono::milliseconds& lookback,
+                                       const std::chrono::milliseconds& counter_history_interval) {
+  std::vector<ChunkInfo> chunk_infos;
+  double sum_hottness = 0.0;
+  size_t lookback_samples = lookback.count() / counter_history_interval.count();
+  for (const auto& table_name : storage_manager.table_names()) {
+    const auto& table = *storage_manager.get_table(table_name);
+    const auto chunk_count = table.chunk_count();
+    for (ChunkID i = ChunkID(0); i < chunk_count; i++) {
+      const auto& chunk = table.get_chunk(i);
+      if (chunk_is_completed(chunk, table.chunk_size()) && chunk.has_access_counter()) {
+        const double hottness = static_cast<double>(chunk.access_counter()->history_sample(lookback_samples));
+        sum_hottness += hottness;
+        chunk_infos.emplace_back(ChunkInfo{.table_name = table_name,
+                                           .id = i,
+                                           .node = get_node_id(chunk.get_allocator()),
+                                           .byte_size = chunk.byte_size(),
+                                           .hottness = hottness});
+      }
+    }
+  }
+  std::sort(chunk_infos.begin(), chunk_infos.end(), [](const ChunkInfo& a, const ChunkInfo& b) { return b < a; });
+  return chunk_infos;
+}
+
+std::vector<size_t> count_chunks_by_node(const std::vector<ChunkInfo>& chunk_infos, size_t node_count) {
+  std::vector<size_t> result(node_count);
```

Comment:
What do we need the `node_count` parameter for if it appears to be equal to `chunk_infos.size()`?